File: packages/engine/src/sourcePack.ts
/// <reference path="./pdf-parse.d.ts" />
import { readFile } from "node:fs/promises";
import { extname } from "node:path";
import mammoth from "mammoth";
// Import the internal entry point to avoid pdf-parse's index.js debug self-test.
import pdfParse from "pdf-parse/lib/pdf-parse.js";

export interface PackedSource {
  id: string;
  label: string;
  kind: "document" | "pdf";
  text?: string; // docx / pdf extracted text
  summary: string;
}

export interface SourcePack {
  sources: PackedSource[];
}

export interface EngineFile {
  id: string;
  name: string;
  mime: string;
  path: string;
}

const MAX_DOCUMENT_CHARS = 8_000;

type SourceKind = "csv" | "xlsx" | "docx" | "pdf" | "unknown";

/** Dispatch on mime type, falling back to file extension. */
function classify(file: EngineFile): SourceKind {
  const mime = file.mime.toLowerCase();
  if (mime.includes("csv")) return "csv";
  if (mime.includes("spreadsheetml") || mime.includes("ms-excel")) return "xlsx";
  if (mime.includes("wordprocessingml") || mime.includes("msword")) return "docx";
  if (mime.includes("pdf")) return "pdf";

  switch (extname(file.name).toLowerCase()) {
    case ".csv":
      return "csv";
    case ".xlsx":
    case ".xls":
      return "xlsx";
    case ".docx":
    case ".doc":
      return "docx";
    case ".pdf":
      return "pdf";
    default:
      return "unknown";
  }
}

/**
 * Build the source pack from document families only — the warehouse owns every
 * tabular file, so csv/xlsx are skipped entirely and never reach the prompt.
 */
export async function buildSourcePack(files: EngineFile[]): Promise<SourcePack> {
  const sources = await Promise.all(
    files.filter((f) => !["csv", "xlsx"].includes(classify(f))).map(buildSource),
  );
  return { sources };
}

function buildSource(file: EngineFile): Promise<PackedSource> {
  switch (classify(file)) {
    case "pdf":
      return buildPdf(file);
    case "docx":
    default:
      // Unknown types fall back to plain-text extraction rather than crashing a run.
      return buildDocument(file);
  }
}

async function buildDocument(file: EngineFile): Promise<PackedSource> {
  const kind = classify(file);
  let text = "";
  if (kind === "docx") {
    const result = await mammoth.extractRawText({ path: file.path });
    text = result.value ?? "";
  } else {
    text = await readFile(file.path, "utf8");
  }
  const words = text.trim() ? text.trim().split(/\s+/).length : 0;
  return {
    id: file.id,
    label: file.name,
    kind: "document",
    text,
    summary: `${file.name} — document, ${words.toLocaleString("en-US")} words`,
  };
}

async function buildPdf(file: EngineFile): Promise<PackedSource> {
  const buffer = await readFile(file.path);
  const kb = Math.max(1, Math.round(buffer.byteLength / 1024));
  // Extract text locally so the PDF's contents flow through the source pack like
  // any document. Never let a malformed PDF crash the whole pack build.
  let text = "";
  try {
    const data = await pdfParse(buffer);
    text = data.text;
  } catch {
    text = "";
  }
  return {
    id: file.id,
    label: file.name,
    kind: "pdf",
    text,
    summary: `${file.name} — PDF, ${kb} KB (read at run time)`,
  };
}

/** Raw extracted text of one file — the same reader buildSourcePack uses, without prompt packaging. */
export async function extractFileText(file: EngineFile): Promise<string> {
  const pack = await buildSourcePack([file]);
  return pack.sources.map((s) => s.text ?? "").join("\n\n");
}

/** Serialize selected sources for a drafting prompt. */
export function packForPrompt(pack: SourcePack, sourceIds: string[]): string {
  const chosen = pack.sources.filter((s) => sourceIds.includes(s.id));
  return chosen.map(serializeSource).join("\n\n");
}

function serializeSource(source: PackedSource): string {
  const parts: string[] = [`### ${source.label} (${source.id})`, source.summary];
  // PDFs are text-extracted locally, so they serialize like documents.
  if (source.text) parts.push(source.text.slice(0, MAX_DOCUMENT_CHARS));
  return parts.join("\n");
}

// --- exceljs cell coercion -------------------------------------------------
// Kept for tabular.ts, which coerces warehouse cell values through cellValue.

export function cellText(value: unknown): string {
  if (value === null || value === undefined) return "";
  if (value instanceof Date) return value.toISOString();
  if (typeof value === "object") {
    const o = value as Record<string, unknown>;
    if (typeof o.text === "string") return o.text;
    if (Array.isArray(o.richText)) return o.richText.map((t: { text?: string }) => t.text ?? "").join("");
    if ("result" in o) return o.result == null ? "" : String(o.result);
    return String(value);
  }
  return String(value);
}

export function cellValue(value: unknown): string | number {
  if (value === null || value === undefined) return "";
  if (typeof value === "number") return value;
  if (typeof value === "boolean") return String(value);
  if (value instanceof Date) return value.toISOString();
  if (typeof value === "object") {
    const o = value as Record<string, unknown>;
    if (typeof o.result === "number") return o.result;
    return cellText(value);
  }
  return String(value);
}

File: backend/runoff_api/engine/source_pack.py
"""Source pack — statement-for-statement port of packages/engine/src/sourcePack.ts.

Builds the document-side prompt surface: the warehouse owns every tabular file,
so csv/xlsx are skipped entirely and never reach the prompt; docx/pdf/unknown
files are text-extracted locally and packaged for drafting prompts.

Runtime shapes are plain dicts with camelCase keys (TS async -> sync here):
  PackedSource  {"id", "label", "kind", "text"?, "summary"}   kind in {"document","pdf"}
  SourcePack    {"sources": [PackedSource, ...]}
  EngineFile    {"id", "name", "mime", "path"}   (exact field set from the TS interface)

Sanctioned near-parity divergence (spec §6): only extract_file_text's PDF branch
is a near-parity path — it uses pypdf, whose text may differ from TS pdf-parse
output. DOCX uses the same mammoth library on both stacks, so it is byte-parity,
not a divergence. Both are fuzzy document text that feeds prompts, never
byte-exact warehouse data. cell_text / cell_value stay byte-exact ports (they
coerce warehouse cell values that feed prompts).
"""

from __future__ import annotations

import datetime as _dt
import os
from typing import Any

import mammoth
import pypdf

from .parse_plan import js_string
from .tabular import _to_iso_string

MAX_DOCUMENT_CHARS = 8_000


def _extname(name: str) -> str:
    return os.path.splitext(name)[1]


def _classify(file: dict) -> str:
    """Dispatch on mime type, falling back to file extension (SourceKind)."""
    mime = file["mime"].lower()
    if "csv" in mime:
        return "csv"
    if "spreadsheetml" in mime or "ms-excel" in mime:
        return "xlsx"
    if "wordprocessingml" in mime or "msword" in mime:
        return "docx"
    if "pdf" in mime:
        return "pdf"

    ext = _extname(file["name"]).lower()
    if ext == ".csv":
        return "csv"
    if ext in (".xlsx", ".xls"):
        return "xlsx"
    if ext in (".docx", ".doc"):
        return "docx"
    if ext == ".pdf":
        return "pdf"
    return "unknown"


def build_source_pack(files: list[dict]) -> dict:
    """Build the source pack from document families only — csv/xlsx are skipped."""
    sources = [
        _build_source(f) for f in files if _classify(f) not in ("csv", "xlsx")
    ]
    return {"sources": sources}


def _build_source(file: dict) -> dict:
    if _classify(file) == "pdf":
        return _build_pdf(file)
    # docx and unknown types fall through to plain-text extraction.
    return _build_document(file)


def _js_thousands(n: int) -> str:
    """Number.toLocaleString('en-US') for a non-negative integer."""
    return f"{n:,}"


def _build_document(file: dict) -> dict:
    kind = _classify(file)
    if kind == "docx":
        text = _extract_docx(file["path"])
    else:
        # Node's readFile(path, "utf8") never throws on invalid bytes (it yields
        # U+FFFD); match that with errors="replace" so a bad byte can't crash the
        # whole pack build.
        with open(file["path"], encoding="utf-8", errors="replace") as fh:
            text = fh.read()
    words = len(text.strip().split()) if text.strip() else 0
    return {
        "id": file["id"],
        "label": file["name"],
        "kind": "document",
        "text": text,
        "summary": f"{file['name']} — document, {_js_thousands(words)} words",
    }


def _build_pdf(file: dict) -> dict:
    size = os.path.getsize(file["path"])
    # Math.round (round half to +inf), min 1.
    kb = max(1, int((size / 1024) + 0.5))
    # Extract text locally so the PDF's contents flow through the source pack like
    # any document. Never let a malformed PDF crash the whole pack build.
    text = ""
    try:
        reader = pypdf.PdfReader(file["path"])
        text = "\n".join(page.extract_text() for page in reader.pages)
    except Exception:
        text = ""
    return {
        "id": file["id"],
        "label": file["name"],
        "kind": "pdf",
        "text": text,
        "summary": f"{file['name']} — PDF, {kb} KB (read at run time)",
    }


def _extract_docx(path: str) -> str:
    """Raw-text extraction from a .docx via python-mammoth — the same library
    (same author) TS uses via mammoth.extractRawText({path}). Python's port takes
    a fileobj; open the path in binary and return the result's `.value`.

    Errors are NOT swallowed: mammoth raises on a missing/corrupt file, matching
    TS. (buildSourcePack only guards the PDF branch, never docx.)
    """
    with open(path, "rb") as fh:
        return mammoth.extract_raw_text(fh).value


def extract_file_text(file: dict) -> str:
    """Raw extracted text of one file — the same reader build_source_pack uses."""
    pack = build_source_pack([file])
    return "\n\n".join(s.get("text") or "" for s in pack["sources"])


def pack_for_prompt(pack: dict, source_ids: list[str]) -> str:
    """Serialize selected sources for a drafting prompt."""
    chosen = [s for s in pack["sources"] if s["id"] in source_ids]
    return "\n\n".join(_serialize_source(s) for s in chosen)


def _serialize_source(source: dict) -> str:
    parts = [f"### {source['label']} ({source['id']})", source["summary"]]
    # PDFs are text-extracted locally, so they serialize like documents.
    text = source.get("text")
    if text:
        parts.append(text[:MAX_DOCUMENT_CHARS])
    return "\n".join(parts)


# --- exceljs cell coercion -------------------------------------------------
# Kept for tabular parity: warehouse cell values coerce through cell_value.
# JS String() semantics come from the shared engine.parse_plan.js_string (the
# None/datetime arms it adds over the old private copy are unreachable here —
# cell_text/cell_value guard both before ever calling it).


def cell_text(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, _dt.datetime):
        return _to_iso_string(value)
    if isinstance(value, dict):
        text = value.get("text")
        if isinstance(text, str):
            return text
        rich = value.get("richText")
        if isinstance(rich, list):
            return "".join((t.get("text") or "") for t in rich)
        if "result" in value:
            result = value.get("result")
            return "" if result is None else js_string(result)
        return js_string(value)
    return js_string(value)


def cell_value(value: Any) -> str | int | float:
    if value is None:
        return ""
    # A JS boolean is not a number; check it before the numeric branch since a
    # Python bool is an int subclass.
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return value
    if isinstance(value, _dt.datetime):
        return _to_iso_string(value)
    if isinstance(value, dict):
        result = value.get("result")
        if isinstance(result, (int, float)) and not isinstance(result, bool):
            return result
        return cell_text(value)
    return js_string(value)
